bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java

```diff
@@ -589,12 +588,15 @@ public void readEntryComplete(int rc, long ledgerId, final long entryId, final B
         heardFromHosts.add(rctx.to);
         heardFromHostsBitSet.set(rctx.bookieIndex, true);
 
+        buffer.retain();
```

Comment:
This is additional retain you added here. Who's responsibility is to release this buffer?

Reply:
We pass it into entry#complete, if entry#complete accepts the buffer, it gets passed to entryImpl.setEntryBuf, which takes ownership, and entry#complete returns true, otherwise it returns false, and it gets cleaned up in the else branch. entryImpls either get passed to the client (which is then responsible for cleanup), or cleaned up in submitCallback L637.